### ISSUE TITLE
Implementation Gap: Current Q-K Projection vs. Atlas Paper's Gradient-Based TTL Memory

### ISSUE THREAD

**Issue by toddwbucy** [toddwbucy/Atlas-MAG_OmegaRule#12]
# Implementation Gap Analysis: Atlas Paper vs. Current Implementation

## Executive Summary

The current implementation is a **closed-form approximation** of the Atlas architecture, not the actual gradient-based test-time learning (TTL) system described in the paper (arXiv:2505.23735). While several key components are implemented and ready, they are not wired together to form the true Atlas algorithm.

**TL;DR**: We have the parts, but they're not connected correctly.

---

### 1.1 The Omega Rule (Equation 9)

The core innovation of Atlas is optimizing memory over a **context window** of tokens, not just the current token:

```
min_M Σ(i=t-c+1 to t) γ_i^(t) ||M(φ(k_i)) - v_i||²_2
```

Where:
- `c` = context window size
- `γ_i^(t)` = input-dependent decay weights
- `M(·)` = deep MLP memory module
- `φ(·)` = polynomial feature mapping

### 1.2 The Atlas Update Rule (Equations 32-33)

```
M_t = α_t * M_{t-1} - η_t * NewtonSchulz-k(S_t)
S_t = θ_t * S_{t-1} + ∇ Σ(i=t-c+1 to t) γ_i^(t) ||M(φ(k_i)) - v_i||²_2
```

Key components:
- **Muon optimizer** (Newton-Schulz iterations) for memory updates
- **Momentum term** S_t with decay θ_t
- **Polynomial feature mapping** φ(·) on keys
- **Context window** of size c (not online/single-token)
- **Input-dependent decay** γ_i^(t)

### 1.3 Memory Architecture (Equations 42-43)

**Standard Atlas Memory:**
```python
M(x) = x + W1 * σ(W2 * x)  # 2-layer MLP with residual
```

**Atlas++ Gated Memory:**
```python
M(x) = x + W1 * (σ(W2(x)) ⊗ W3(x))  # Gated MLP
```

### 1.4 MAG Architecture (Figure 3 in paper)

The paper's MAG (Memory-Attention-Gated) architecture shows:

```
Input → RMSNorm → [Atlas Layer] ──┐
                                  ├──► Add → MLP → Output
              → [SWA] ───────────┘
```

**Critical**: Atlas layer and SWA are **parallel branches that get ADDED**, not a Q-modification followed by attention.

### 1.5 Table 1 Summary (from paper)

| Model | Attentional Bias | Optimizer | Memory Write Operation |
|-------|-----------------|-----------|----------------------|
| **Atlas** | `Σ γ_i ||M(φ(k_i)) - v_i||²` | **Muon** | `M_t = α_t*M_{t-1} - η_t*NS-5(S_t)` |

---

### 2.1 Why Polynomial Features Are Required

The Atlas paper explicitly recommends polynomial feature mapping to address **limited memory capacity**. This is not a performance optimization—it's compensating for a fundamental architectural bottleneck.

**From Section 3.1 of the paper:**
> "To improve the limited memory capacity, we suggest using higher-order feature mappings (e.g., polynomial feature kernels) on input tokens. We provide theoretical justifications on why deeper memory modules and/or higher-order feature mapping can enhance memory capacity—i.e., the maximum number of linearly independent key-value associations the memory can perfectly map."

### 2.2 Mathematical Justification

**Proposition 1 (Capacity WITHOUT polynomial features):**
```
Let M be a matrix-valued memory with d_v × d_k parameters that optimizes:
    ℓ(M_t; k_t, v_t) = ||M_t * k_t - v_t||²_2

M can store at most O(d_k) pairs of (k_i, v_i) with linearly independent keys.
```

**Proposition 2 (Capacity WITH polynomial features):**
```
Let φ_p(·) be a polynomial mapping with degree at most p.
Let M optimize:
    ℓ(M_t; φ_p(k_t), v_t) = ||M_t * φ_p(k_t) - v_t||²_2

M can store at most O(d_k^p) pairs of (k_i, v_i) with linearly independent keys.
```

### 2.3 The Polynomial Feature Mapping (Equation 22)

```
φ_p(x) = x^⊗p = x ⊗ x^⊗(p-1)    (self-tensoring via Kronecker product)

Dimension of lifted feature space:
D = Σ(ℓ=0 to p) C(d_k + ℓ - 1, ℓ) = C(d_k + p, p)
```

For degree p=2 and d_k=64:
- Without φ: capacity = O(64) = 64 associations
- With φ_2: capacity = O(64²) = 4,096 associations (64× improvement)

### 2.4 Empirical Validation (Paper Table 6 - Ablation Study)

| Model | Perplexity ↓ | Accuracy ↑ |
|-------|-------------|-----------|
| Atlas | 19.97 | 52.77% |
| **w/o Polynomial Mapping** | **22.14** | **50.57%** |

**Impact: +2.17 perplexity, -2.2% accuracy** — a significant degradation.

### 2.5 Critical Consideration for Independent Researchers

The Atlas authors (Google Research) had access to high-end datacenter GPUs/TPUs and still recommended polynomial features for memory capacity reasons. 

**For independent researchers with constrained hardware:**
- Smaller models → smaller d_k dimensions
- Smaller d_k → even more severe capacity bottleneck
- **Polynomial features are MORE critical at smaller scales, not less**

| Context | d_k | Capacity w/o φ | Capacity w/ φ_2 | Improvement |
|---------|-----|----------------|-----------------|-------------|
| Google (large) | 768 | 768 | 589,824 | 768× |
| Independent (small) | 384 | 384 | 147,456 | 384× |
| Tiny | 64 | 64 | 4,096 | 64× |

**Conclusion: Polynomial features must be treated as ESSENTIAL, not optional.**

---

### 3.1 Current Architecture Flow

```
h = norm(x)
q, k, v = QKV(h)
γ_gates = GammaGate(h)                    # ✅ Input-dependent decay
q_projected = CausalQKMemoryProjection(q, k, γ)   # ❌ Outer product, not MLP
q = (1-gate)*q + gate*q_projected          # ❌ Q-level blending, not output-level
attn_out = softmax(q @ k^T / √d) @ v       # Standard attention with modified Q
```

### 3.2 Current "Memory" Computation (qk_projection.py)

The current implementation computes:
```
M_t = M_persistent + Σ(i=t-c+1 to t) γ^(t-i) * gate_i * (k_i ⊗ k_i)
q'_t = M_t @ q_t / norm_sum_t
```

This is a **closed-form approximation** of one gradient step on the ℓ2 loss, using:
- Outer products `k_i ⊗ k_i` instead of MLP parameters
- Direct accumulation instead of gradient descent
- Linear projection instead of MLP forward pass

### 3.3 Current Gate Blending

```python
# skeleton.py:261-263
gate = torch.sigmoid(self.memory_gate)
q = (1 - gate) * q + gate * q_projected
```

The gate blends **at the Q level** (before attention), rather than combining attention output with memory output.

---

## 4. Detailed Gap Analysis

| Aspect | Atlas Paper | Current Implementation | Gap |
|--------|-------------|----------------------|-----|
| **Memory Module M** | Deep MLP with residual: `M(x) = x + W1*σ(W2*x)` | Outer product sum: `Σ k_i ⊗ k_i` | **Architectural** |
| **Memory Update** | Gradient descent: `M_t = α*M_{t-1} - η*∇loss` | Closed-form accumulation | **Algorithmic** |
| **Optimizer** | Muon (Newton-Schulz on momentum) | Not used for memory | **Algorithmic** |
| **Momentum** | `S_t = θ*S_{t-1} - ∇loss` | Not implemented | **Missing** |
| **Memory Output** | `M(φ(k_t))` - MLP forward pass | `M_t @ q / norm` - linear projection | **Architectural** |
| **Combination Point** | Output level: `attn_out + mem_out` | Q level: `(1-g)*q + g*q'` | **Architectural** |
| **Polynomial Features** | `φ(k)` applied to keys (**ESSENTIAL**) | Implemented but unused | **Wiring** |
| **Context Window** | Optimizes over window of c tokens | ✅ Implemented correctly | OK |
| **Input-dependent γ** | Per-position decay gates | ✅ Implemented correctly | OK |

---

### 5.1 AtlasMemory (atlas_memory.py:29-98) ✅ READY

```python
class AtlasMemory(nn.Module):
    """Gated MLP memory - matches paper's Eq. 43"""
    def forward(self, x):
        gate = F.silu(self.w2(x))
        value = self.w3(x)
        gated = gate * value
        out = x + self.w1(gated)  # Residual
        return out
```

**Status**: Implemented, matches paper, NOT USED in main forward pass.

### 5.2 AtlasMemoryPoly (atlas_memory.py:105-225) ✅ READY — ESSENTIAL

```python
class AtlasMemoryPoly(nn.Module):
    """Memory with degree-2 polynomial feature mapping φ_p(x)"""
    def _polynomial_features(self, x):
        # Implements φ_2(x) = [x, x⊗x] (upper triangular)
        outer = x.unsqueeze(-1) * x.unsqueeze(-2)
        triu = outer[..., triu_idx[0], triu_idx[1]]
        return torch.cat([x, triu], dim=-1)
```

**Status**: Implemented, NOT USED. Config has `POLY_DEGREE=2` but not wired in.

**⚠️ ESSENTIAL**: Per paper's Proposition 2, this transforms capacity from O(d_k) to O(d_k²).

### 5.3 Newton-Schulz / Muon (newton_schulz.py:24-65) ✅ READY

```python
def newton_schulz(G: Tensor, num_iters: int = 5, eps: float = 1e-7) -> Tensor:
    """Orthogonalize matrix via Newton-Schulz iteration."""
    X = G / (norm + eps)
    for _ in range(num_iters):
        A = X @ X.transpose(-2, -1)
        B = _NS_B * A + _NS_C * (A @ A)  # Optimal coefficients
        X = _NS_A * X + B @ X
    return X
```

**Status**: Correctly implemented with optimal coefficients (a=3.4445, b=-4.7750, c=2.0315), NOT USED for memory updates.

### 5.4 GammaGate (skeleton.py:39-86) ✅ USED

```python
class GammaGate(nn.Module):
    """Input-dependent decay gates"""
    # x -> Linear -> SiLU -> Linear -> Sigmoid -> γ
```

**Status**: Implemented and correctly integrated.

---

### 6.1 Gradient Computation for Memory Update ❌ NOT IMPLEMENTED

Need to compute:
```python
# Omega loss over context window
loss = sum(γ_i * ||M(φ(k_i)) - v_i||² for i in range(t-c+1, t+1))

# Gradient w.r.t. M's parameters
grad = torch.autograd.grad(loss, M.parameters())
```

### 6.2 Momentum Accumulation ❌ NOT IMPLEMENTED

```python
# Paper Eq. 33
S_t = θ_t * S_{t-1} - gradient
```

Need to maintain momentum state across forward passes.

### 6.3 Memory Update Loop ❌ NOT IMPLEMENTED

```python
# Paper Eq. 32
S_t_orthogonalized = newton_schulz(S_t, num_iters=5)
for param, grad in zip(M.parameters(), S_t_orthogonalized):
    param.data = α_t * param.data - η_t * grad
```

### 6.4 Output-Level Combination ❌ NOT IMPLEMENTED

Current: Q-level blending before attention
Needed: Output-level combination after both attention and memory

```python
# Paper's MAG architecture
attn_out = attention(q, k, v)
mem_out = M(φ(k))  # or M(x)
combined = attn_out + mem_out  # Or gated combination
```

---

### What Works in Current Implementation
- ✅ Basic architecture skeleton (blocks, norms, FFN)
- ✅ Input-dependent γ gates
- ✅ Context window concept (sliding window)
- ✅ Persistent memory precomputation
- ✅ Gate polarization loss (training innovation)
- ✅ Per-layer gate initialization

### What Doesn't Match Paper
- ❌ No gradient-based memory updates (core TTL innovation)
- ❌ No deep MLP memory (uses linear outer products)
- ❌ No Muon optimizer for memory
- ❌ No momentum term
- ❌ Wrong combination point (Q vs output)
- ❌ Polynomial features unused (**essential for capacity**)

### Severity
**HIGH** - The core innovation of Atlas (test-time learning via gradient-based memory updates) is not implemented. The current implementation is a different (simpler) architecture that approximates some behaviors but lacks the key capability.

---

### Branch 1: `refactor/output-level-combination`
- Move combination from Q-level to output-level
- Implement parallel branches (Atlas ‖ SWA → Add)
- **No dependencies**

### Branch 2: `feature/wire-atlas-memory`
- Wire `AtlasMemory` into forward pass
- Wire `AtlasMemoryPoly` with polynomial features φ(k) (**ESSENTIAL**)
- Add outer-loop weight updates (standard backprop)
- **Depends on**: Branch 1

### Branch 3: `feature/ttl-update-loop` (THE BIG ONE)
- Implement gradient computation through memory
- Implement momentum accumulation (S_t)
- Wire Newton-Schulz into memory update
- Implement full update rule: `M_t = α*M_{t-1} - η*NS(S_t)`
- **Depends on**: Branch 2
- **Note**: This is a monolithic unit—cannot be split

### Branch 4: `feature/training-integration`
- Handle memory state across batches
- Update gate monitoring
- Adjust polarization if needed
- **Depends on**: Branch 3

---

## 9. Files Affected

| File | Changes Needed |
|------|---------------|
| `src/model/skeleton.py` | Rewrite AtlasMAGBlock to use AtlasMemory, change combination point |
| `src/model/atlas_memory.py` | Add momentum state, gradient computation |
| `src/model/qk_projection.py` | May be deprecated or repurposed |
| `src/model/newton_schulz.py` | Wire into memory update |
| `scripts/train_smollm.py` | Handle memory state across batches |
| `src/config.py` | Add Muon/momentum hyperparameters |

---

## 10. References

- **Atlas Paper**: arXiv:2505.23735 "Atlas: Learning to Optimally Memorize the Context at Test Time"
- **Key Equations**: 9 (Omega Rule), 32-33 (Update), 42-43 (Memory Architecture)
- **Key Figures**: Figure 3 (MAG Architecture), Table 1 (Model Comparison)
- **Capacity Proofs**: Propositions 1 & 2 (Section 3.1)
- **Ablation Study**: Table 6 (Section 6.6)

---

## 11. Acceptance Criteria

- [ ] `AtlasMemory` used as the memory module M(·)
- [ ] `AtlasMemoryPoly` with polynomial features φ(k) enabled (**ESSENTIAL**)
- [ ] Omega loss computed over context window
- [ ] Gradients computed w.r.t. M's parameters
- [ ] Newton-Schulz applied to momentum term
- [ ] Memory updated via: `M_t = α*M_{t-1} - η*NS(S_t)`
- [ ] Combination at output level (not Q level)
- [ ] Tests verify gradient flow through memory
- [ ] Training converges comparably to paper
